addons/mail_group/models/mail_group.py

```diff
@@ -0,0 +1,319 @@
+        help='Newcomers on this moderated group will automatically receive the guidelines.')
+    moderation_guidelines_msg = fields.Text(string='Guidelines')
+
+    @api.depends('mail_group_message_ids')
+    def _compute_mail_group_message_count(self):
+        for group in self:
+            group.mail_group_message_count = len(group.mail_group_message_ids.filtered(
+                lambda message: message.moderation_status == 'pending_moderation'
+            ))
+
+    @api.depends('member_ids')
+    def _compute_member_count(self):
+        for group in self:
+            group.member_count = len(group.member_ids)
+
+    @api.depends('moderator_ids')
+    @api.depends_context('uid')
+    def _compute_is_moderator(self):
+        for group in self:
+            group.is_moderator = self.env.user.id in group.moderator_ids.ids
+
+    @api.depends('moderation_ids')
+    def _compute_moderation_count(self):
+        for group in self:
+            group.moderation_count = len(group.moderation_ids)
+
+    def _alias_get_creation_values(self):
+        """Return the default values for the automatically created alias."""
+        values = super(MailGroup, self)._alias_get_creation_values()
+        values['alias_model_id'] = self.env['ir.model']._get('mail.group').id
+        values['alias_force_thread_id'] = self.id
+        values['alias_defaults'] = literal_eval(self.alias_defaults or '{}')
+        return values
+
+    @api.onchange('moderation')
+    def _onchange_moderation(self):
+        if self.moderation and self.env.user not in self.moderator_ids:
+            self.moderator_ids |= self.env.user
+
+    # CONSTRAINTS
+
+    @api.constrains('moderator_ids')
+    def _check_moderator_email(self):
+        if any(not moderator.email for group in self for moderator in group.moderator_ids):
+            raise ValidationError(_('Moderators must have an email address.'))
+
+    @api.constrains('moderation_notify', 'moderation_notify_msg')
+    def _check_moderation_notify(self):
+        if any(group.moderation_notify and not group.moderation_notify_msg for group in self):
+            raise ValidationError(_('The notification message is missing.'))
+
+    @api.constrains('moderation_guidelines', 'moderation_guidelines_msg')
+    def _check_moderation_guidelines(self):
+        if any(group.moderation_guidelines and not group.moderation_guidelines_msg for group in self):
+            raise ValidationError(_('The guidelines description is missing.'))
+
+    @api.constrains('moderator_ids', 'moderation')
+    def _check_moderator_existence(self):
+        if any(not group.moderator_ids for group in self if group.moderation):
+            raise ValidationError(_('Moderated group must have moderators.'))
+
+    @api.model
+    def message_update(self, message_dict=None):
+        """Add the method to make the mail gateway flow work with this model."""
+        return
+
+    def message_post(self, **kwargs):
```

Comment:
We could also perform some checks on parent_id (if given, checks it belong to the right model/res_id), just to be sure values given make sense.

Reply:
I added a constraint about that `_check_mail_message_id` which check both model and res_id